lib/proxy/clientmetrics.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// clientMetrics represents a collection of metrics for a proxy peer client
+type clientMetrics struct {
+	tunnelErrorCounter *prometheus.CounterVec
+
+	connections     *prometheus.GaugeVec
+	rpcs            *prometheus.GaugeVec
+	rpcTotal        *prometheus.CounterVec
+	rpcDuration     *prometheus.HistogramVec
+	messageSent     *prometheus.HistogramVec
+	messageReceived *prometheus.HistogramVec
+}
+
+// newClientMetrics inits and registers client metrics prometheus collectors.
+func newClientMetrics() (*clientMetrics, error) {
+	cm := &clientMetrics{
+		tunnelErrorCounter: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "proxy_peer",
+				Subsystem: "client",
+				Name:      "dial_error_total",
+				Help:      "Total number of errors encountered dialing peer proxies.",
+			},
+			[]string{"error_type"},
+		),
+
+		connections: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: "proxy_peer",
+				Subsystem: "client",
+				Name:      "connections",
+				Help:      "Number of currently opened connection to proxy peer servers.",
+			},
+			[]string{"local_addr", "remote_addr"},
+		),
+
+		rpcs: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: "proxy_peer",
+				Subsystem: "client",
+				Name:      "rpc",
+				Help:      "Number of current client RPC requests.",
+			},
+			[]string{"service", "method"},
+		),
+
+		rpcTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "proxy_peer",
+				Subsystem: "client",
+				Name:      "rpc_total",
+				Help:      "Total number of client RPC requests.",
+			},
+			[]string{"service", "method", "code"},
+		),
+
+		rpcDuration: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "proxy_peer",
+				Subsystem: "client",
+				Name:      "rpc_duration_seconds",
+				Help:      "Duration in seconds of RPCs sent by the client.",
+			},
+			[]string{"service", "handler", "code"},
+		),
+
+		messageSent: prometheus.NewHistogramVec(
```

Comment:
This goes for both the client and server side. Why did you choose a histogram over a counter for `messageSent` and `messageReceived`?

Reply:
Ah the variable name relays the wrong info. We're not counting the number of messages sent and received, I don't think we care about that. We're just reporting the size. Unless you meant why we're not reporting the size in a counter in your question?

Reply:
> Unless you meant why we're not reporting the size in a counter in your question?

This is what I meant. Whats the benefits of a histogram vs counter here?

Reply:
I can see this working ok with a counter, but isn't this kind of data more suitable for a histogram?

Reply:
> the total sum of all observed values, exposed as <basename>_sum
Ah I didn't know that histograms also come with a sum. So that works for me